### PR TITLE
Initial version of A1 interface

### PR DESCRIPTION
**Pull request by adibrastegarnia**
This PR adds an initial version of A1 interface based on gRPC and also an auto generated reverse proxy (the focus is not on this part yet and we can use other ways to generate it) based on a given configuration. The terminology which is used in defining of the API is based on the spec. The idea of using gRPC instead of REST API and use a reverse proxy for the applications that want to use REST API instead of gRPC is a valid solution and some of the implementations also introduced a mediator or  adapter layer to expose a set of high level functions to the consumers of the API which tries to map the A1 server functions to the exposed functions (e.g. https://wiki.onap.org/display/DW/A1+Adapter+in+ONAP) and (https://github.com/o-ran-sc/ric-plt-a1).

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -49,7 +49,7 @@ protos: # @HELP compile the protobuf files (using protoc-go Docker)
 	docker run -it -v `pwd`:/go/src/github.com/onosproject/onos-ric \
 		-w /go/src/github.com/onosproject/onos-ric \
 		--entrypoint build/bin/compile-protos.sh \
-		onosproject/protoc-go:${ONOS_PROTOC_VERSION}
+		onosproject/protoc-go:${ONOS_RIC_VERSION}
```
Comment (adibrastegarnia) [inline Makefile:52]: It should be changed before merge.

Comment (adibrastegarnia) [reply Makefile]: We should merge this PR and release the proto image: https://github.com/onosproject/build-tools/pull/37 and then update the ONOS_PROTOC_VERSION in this PR.

File: api/nb/a1/a1.proto [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    string detail = 2;
+    // TODO some other fields should be added
+}
+
+
+
+// PolicyStatement 	Expression of a directive in an A1 policy
+// that is related to policy objectives and/or policy resources
+// and are to be applied to/for the entities identified by the scope identifier.
+message PolicyStatement {
```
Comment (adibrastegarnia) [inline api/nb/a1/a1.proto:37]: I should think about this message a little bit more later.

File: api/nb/a1/a1-p/tsp/tsp.proto [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+enum Preference {
+    SHALL = 0;
+    PREFER = 1;
+    AVOID = 2;
+    FORBID = 3;
+}
+
+// TspResources Attributes used to schedule traffic
+// on available cells in a different way than what would be through default behavior
+message  TspResources {
```
Comment (SeanCondon) [inline api/nb/a1/a1-p/tsp/tsp.proto:33]: I presume TSP is Traffic Steering ... - might be no harm to say it somewhere

File: api/nb/a1/a1.proto [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    rpc CreateOrUpdate (CreateOrUpdateRequest) returns (CreateOrUpdateResponse);
+
+    // Query queries about one or more than one A1 policies
+    rpc Query (QueryRequest) returns (stream QueryResponse);
+    
+    // Delete deletes an A1 policy
+    rpc Delete (DeleteRequest) returns (DeleteResponse);
+
+    // Notify notify about an enforcement status change of a policy between 'enforced' and 'not enforced'.
+    rpc Notify (stream NotifyRequest) returns (stream NotifyResponse);
```
Comment (SeanCondon) [inline api/nb/a1/a1.proto:147]: does it really need to be a bidirectional stream?

File: api/nb/a1/a1_config.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+type: google.api.Service
```
Comment (SeanCondon) [inline api/nb/a1/a1_config.yaml:1]: nice - this creates the REST interface I presume?

File: api/nb/a1/types/types.proto [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+package a1.types;
```
Comment (SeanCondon) [inline api/nb/a1/types/types.proto:19]: does it really need to be put in a types folder? it is called types afterall

File: api/nb/a1/a1-p/qos/qos.proto [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
```
Comment (SeanCondon) [inline api/nb/a1/a1-p/qos/qos.proto:18]: does this really need to be nested inside the a1-p folder? Same with next one

File: api/nb/a1/a1.proto (hunk 4 above)
Comment (adibrastegarnia) [reply api/nb/a1/a1.proto:147]: Not sure yet. Perhaps server side streaming should be enough. We can change it later if it is needed

File: api/nb/a1/a1_config.yaml (hunk 5 above)
Comment (adibrastegarnia) [reply api/nb/a1/a1_config.yaml:1]: There are two ways to annotate a protobuf for generating a reverse proxy. One approach is to annotate the protobuf itself and another one is to define a config file which maps rpcs and their fields to REST API resources. Haven't tested yet but just wanted to make sure the mapping is possible based on definition of API in gRPC. REST API reverse proxy is generated also which is named a1.pb.gw.go

File: api/nb/a1/types/types.proto (hunk 6 above)
Comment (adibrastegarnia) [reply api/nb/a1/types/types.proto:19]: Perhaps not for now. I will take it out.

File: api/nb/a1/a1-p/qos/qos.proto (hunk 7 above)
Comment (adibrastegarnia) [reply api/nb/a1/a1-p/qos/qos.proto:18]: a1 has multiple models. one of them is a1-p. there are some other models which are not defined yet. qos and tsp are under a1-p model. I wanted to make sure they are organized in a meaningful way.

File: api/nb/a1/a1-p/tsp/tsp.proto (hunk 3 above)
Comment (adibrastegarnia) [reply api/nb/a1/a1-p/tsp/tsp.proto:33]: Will take care of it.

**Pull request merged**